mistralrs-core/src/sampler.rs

```diff
@@ -379,8 +488,71 @@ impl Sampler {
                     - if count > 0.0 { 1. } else { 0. } * presence_penalty;
             }
         }
-        let vocab_size = logits.len();
-        Tensor::from_vec(logits, vocab_size, &Device::Cpu)
+        Ok(())
+    }
+
+    fn apply_dry_penalty(&self, logits: &mut [f32], context: &[u32]) -> Result<()> {
+        if let Some(ref params) = self.dry_params {
+            let match_indices = context
+                .par_iter()
+                .enumerate()
+                .take(context.len() - 1)
+                .filter(|(_i, x)| *context.last().unwrap() == **x)
+                .map(|(i, _)| i)
+                .collect::<Vec<_>>();
+
+            let mut match_lengths = HashMap::new();
+
+            for i in match_indices {
+                let next_token = context[i + 1];
+
+                if params.sequence_breakers.contains(&next_token) {
+                    continue;
+                }
+
+                let mut match_length = 1;
+
+                // Limit match length to avoid quadratic runtime and potential DoS with adversarial inputs.
+                while match_length < 50 {
+                    if match_length > i {
+                        // Start of input
+                        break;
+                    }
+
+                    let j = i - match_length;
+
+                    let prev_tok = context[context.len() - (match_length + 1)];
+                    if context[j] != prev_tok {
+                        // Start of match reached
+                        break;
+                    }
+
+                    if params.sequence_breakers.contains(&prev_tok) {
+                        // Seq breaking tok reached
+                        break;
+                    }
+
+                    match_length += 1;
+                }
+
+                #[allow(clippy::map_entry)]
+                if match_lengths.contains_key(&next_token) {
+                    match_lengths.insert(next_token, match_length.max(match_lengths[&next_token]));
+                } else {
+                    match_lengths.insert(next_token, match_length);
+                }
+            }
+
+            // Actually apply penalties
+            for (tok, match_len) in match_lengths {
+                if match_len >= params.allowed_length {
+                    let penalty = params.multiplier
+                        * params.base.powf((match_len - params.allowed_length) as f32);
+                    logits[tok as usize] = penalty;
```

Comment:
```suggestion
                    logits[tok as usize] -= penalty;
```

@EricLBuehler 

:rocket: :exclamation: :exclamation: :exclamation: :exclamation: :exclamation: :rocket: 

Finally, I have found the devil, after looking under every rock! I must have read past that line half a dozen times while debugging, thinking "it can't be there", before noticing the problem. :stuck_out_tongue: 

I'm still not 100% certain the implementation is correct, but the artifacts are all gone now, and both with and without DRY the output looks OK.

Reply:
@p-e-w thanks for finding this!! This sounds good, please feel free to approve the PR or do any additional reviews. I'll do some more testing on my end and hopefully we can merge this today.